images/virtualization-artifact/pkg/controller/gc/cron_source.go

```diff
@@ -17,157 +17,92 @@ limitations under the License.
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	"github.com/deckhouse/deckhouse/pkg/log"
-	"github.com/deckhouse/virtualization-controller/pkg/common/object"
 	"github.com/deckhouse/virtualization-controller/pkg/logger"
 )
 
-const sourceName = "CronSource"
-
-func NewCronSource(c client.Client,
-	standardSpec string,
-	objList client.ObjectList,
-	option CronSourceOption,
-	log *log.Logger,
-) *CronSource {
-	return &CronSource{
-		Client:       c,
-		standardSpec: standardSpec,
-		objList:      objList,
-		option:       option,
-		log:          log.With("WatchSource", sourceName),
-	}
-}
-
 var _ source.Source = &CronSource{}
 
-type CronSource struct {
-	client.Client
-	standardSpec string
-	objList      client.ObjectList
-	option       CronSourceOption
-	log          *log.Logger
-}
+const sourceName = "CronSource"
 
-type CronSourceOption struct {
-	GetOlder func(objList client.ObjectList) client.ObjectList
+type SourceGCManager interface {
+	ListForDelete(ctx context.Context, now time.Time) ([]client.Object, error)
 }
 
-func NewDefaultCronSourceOption(objs client.ObjectList, ttl time.Duration, log *log.Logger) CronSourceOption {
-	return CronSourceOption{
-		GetOlder: DefaultGetOlder(objs, ttl, 10, log),
+func NewCronSource(scheduleSpec string, client client.Client, mgr SourceGCManager, log *log.Logger) (*CronSource, error) {
+	schedule, err := cron.ParseStandard(scheduleSpec)
+	if err != nil {
+		return nil, fmt.Errorf("parsing standard spec %q: %w", scheduleSpec, err)
 	}
-}
 
-func DefaultGetOlder(objs client.ObjectList, ttl time.Duration, maxCount int, log *log.Logger) func(objList client.ObjectList) client.ObjectList {
-	return func(objList client.ObjectList) client.ObjectList {
-		var expiredItems []runtime.Object
-		var notExpiredItems []runtime.Object
-
-		if err := meta.EachListItem(objList, func(o runtime.Object) error {
-			obj, ok := o.(client.Object)
-			if !ok {
-				return nil
-			}
-			if object.GetAge(obj) > ttl {
-				expiredItems = append(expiredItems, o)
-			} else {
-				notExpiredItems = append(notExpiredItems, o)
-			}
-
-			return nil
-		}); err != nil {
-			log.Error("failed to populate list", logger.SlogErr(err))
-		}
-
-		if maxCount != 0 && len(notExpiredItems) > maxCount {
-			slices.SortFunc(notExpiredItems, func(a, b runtime.Object) int {
-				aObj, _ := a.(client.Object)
-				bObj, _ := b.(client.Object)
-
-				return cmp.Compare(object.GetAge(aObj), object.GetAge(bObj))
-			})
-			expiredItems = append(expiredItems, notExpiredItems[maxCount:]...)
-		}
+	return &CronSource{
+		schedule: schedule,
+		client:   client,
+		mgr:      mgr,
+		log:      log.With("WatchSource", sourceName),
+		clock:    &clock.RealClock{},
+	}, nil
+}
 
-		if err := meta.SetList(objs, expiredItems); err != nil {
-			log.Error("failed to set list", logger.SlogErr(err))
-		}
-		return objs
-	}
+type CronSource struct {
+	schedule cron.Schedule
+	client   client.Client
```

Comment:
Is `client` needed? It seems CrontSource do not call the client. Everything `client`-related is hidden into `mgr`, so client is not needed anymore.